cmd/snap-confine/udev-support.c

```diff
@@ -18,301 +18,403 @@
+	}
+}
 
-	if (udev_enumerate_scan_devices(udev_s->devices) < 0)
-		die("udev_enumerate_scan failed");
+static void sc_udev_setup_acls(int devices_allow_fd, int devices_deny_fd,
+			       struct udev *udev,
+			       struct udev_list_entry *assigned)
+{
+	/* Deny device access by default.
+	 *
+	 * Write 'a' to devices.deny to remove all existing devices that were added
+	 * in previous launcher invocations, then add the static and assigned
+	 * devices. This ensures that at application launch the cgroup only has
+	 * what is currently assigned. */
+	sc_dprintf(devices_deny_fd, "a");
 
-	udev_s->assigned = udev_enumerate_get_list_entry(udev_s->devices);
-	if (udev_s->assigned == NULL)
-		rc = -1;
+	/* Allow access to various devices. */
+	sc_udev_allow_common(devices_allow_fd);
+	sc_udev_allow_pty_slaves(devices_allow_fd);
+	sc_udev_allow_nvidia(devices_allow_fd);
+	sc_udev_allow_uhid(devices_allow_fd);
+	sc_udev_allow_dev_net_tun(devices_allow_fd);
+	sc_udev_allow_assigned(devices_allow_fd, udev, assigned);
+}
 
-	return rc;
+static char *sc_udev_mangle_security_tag(const char *security_tag)
+{
+	char *udev_tag = sc_strdup(security_tag);
+	for (char *c = strchr(udev_tag, '.'); c != NULL; c = strchr(c, '.')) {
+		*c = '_';
+	}
+	return udev_tag;
 }
 
-void snappy_udev_cleanup(struct snappy_udev *udev_s)
+static void sc_cleanup_udev(struct udev **udev)
 {
-	// udev_s->assigned does not need to be unreferenced since it is a
-	// pointer into udev_s->devices
-	if (udev_s->devices != NULL)
-		udev_enumerate_unref(udev_s->devices);
-	if (udev_s->udev != NULL)
-		udev_unref(udev_s->udev);
+	if (udev != NULL && *udev != NULL) {
+		udev_unref(*udev);
+		*udev = NULL;
+	}
 }
 
-void setup_devices_cgroup(const char *security_tag, struct snappy_udev *udev_s)
+static void sc_cleanup_udev_enumerate(struct udev_enumerate **enumerate)
 {
-	debug("%s", __func__);
-	// Devices that must always be present
-	const char *static_devices[] = {
-		"/sys/class/mem/null",
-		"/sys/class/mem/full",
-		"/sys/class/mem/zero",
-		"/sys/class/mem/random",
-		"/sys/class/mem/urandom",
-		"/sys/class/tty/tty",
-		"/sys/class/tty/console",
-		"/sys/class/tty/ptmx",
-		NULL,
-	};
-
-	if (udev_s == NULL)
-		die("snappy_udev is NULL");
-	if (udev_s->udev == NULL)
-		die("snappy_udev->udev is NULL");
-	if (udev_s->devices == NULL)
-		die("snappy_udev->devices is NULL");
-	if (udev_s->assigned == NULL)
-		die("snappy_udev->assigned is NULL");
-	if (udev_s->tagname_len == 0
-	    || udev_s->tagname_len >= MAX_BUF
-	    || strnlen(udev_s->tagname, MAX_BUF) != udev_s->tagname_len
-	    || udev_s->tagname[udev_s->tagname_len] != '\0')
-		die("snappy_udev->tagname has invalid length");
-
-	// create devices cgroup controller
-	char cgroup_dir[PATH_MAX] = { 0 };
-
-	sc_must_snprintf(cgroup_dir, sizeof(cgroup_dir),
-			 "/sys/fs/cgroup/devices/%s/", security_tag);
+	if (enumerate != NULL && *enumerate != NULL) {
+		udev_enumerate_unref(*enumerate);
+		*enumerate = NULL;
```

Comment:
The man page for `udev_enumerate_new` doesn't mention anything about opaqueness like the one for `udev_new` did, but the udev_enumerate_unref() should be sufficient AIUI (which is why the old code also didn't set the pointer to NULL. Unless this is fixing an observed problem (eg, are you doing this cause of valgrind?), I think I'd prefer to stay with the old method (I'm not sure that setting this to NULL isn't breaking the opaqueness rule of what is returned by udev_new since in order to allocate the udev_enumerate, we must call udev_enumerate_new() passing in the udev struct from udev_new()).

Reply:
My comment above for `sc_cleanup_udev` stands. This is the double pointer cleanup function.